sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java

```diff
@@ -113,6 +153,302 @@ public void validateWindow(SqlNode windowOrId, SqlValidatorScope scope, @Nullabl
+  {
+    final DruidSqlIngest ingestNode = (DruidSqlIngest) insert;
+    if (insert.isUpsert()) {
+      throw InvalidSqlInput.exception("UPSERT is not supported.");
+    }
+
+
+    // SQL-style INSERT INTO dst (a, b, c) is not (yet) supported.
+    final String operationName = insert.getOperator().getName();
+    if (insert.getTargetColumnList() != null) {
+      throw InvalidSqlInput.exception(
+          "Operation [%s] cannot be run with a target column list, given [%s (%s)]",
+          operationName,
+          ingestNode.getTargetTable(), ingestNode.getTargetColumnList()
+      );
+    }
+
+    // The target namespace is both the target table ID and the row type for that table.
+    final SqlValidatorNamespace targetNamespace = getNamespace(insert);
+    final IdentifierNamespace insertNs = (IdentifierNamespace) targetNamespace;
+    // The target is a new or existing datasource.
+    final DatasourceTable table = validateInsertTarget(targetNamespace, insertNs, operationName);
+
+    // An existing datasource may have metadata.
+    final DatasourceFacade tableMetadata = table == null ? null : table.effectiveMetadata().catalogMetadata();
+
+    // Validate segment granularity, which depends on nothing else.
+    if (!(ingestNode.getTargetTable() instanceof ExternalDestinationSqlIdentifier)) {
+      Granularity effectiveGranularity = getEffectiveGranularity(operationName, ingestNode, tableMetadata);
+      // Note: though this is the validator, we cheat a bit and write the target
+      // granularity into the query context. Perhaps this step should be done
+      // during conversion, however, we've just worked out the granularity, so we
+      // do it here instead.
+      try {
+        plannerContext.queryContextMap().put(
+            DruidSqlInsert.SQL_INSERT_SEGMENT_GRANULARITY,
+            plannerContext.getPlannerToolbox().jsonMapper().writeValueAsString(effectiveGranularity)
+        );
+      }
+      catch (JsonProcessingException e) {
+        throw InvalidSqlInput.exception(e, "Invalid partition granularity [%s]", effectiveGranularity);
+      }
+    }
+
+    // The source must be a SELECT
+    final SqlNode source = insert.getSource();
+
+    // Validate the source statement.
+    // Because of the non-standard Druid semantics, we can't define the target type: we don't know
+    // the target columns yet, and we can't infer types when they must come from the SELECT.
+    // Normally, the target type is known, and is pushed into the SELECT. In Druid, the SELECT
+    // usually defines the target types, unless the catalog says otherwise. Since catalog entries
+    // are optional, we don't know the target type until we validate the SELECT. (Also, we won't
+    // know names and we match by name.) Thus, we'd have to validate (to know names and types)
+    // to get the target types, but we need the target types to validate. Catch-22. So, we punt.
+    final SqlValidatorScope scope;
+    if (source instanceof SqlSelect) {
```

Comment:
note: it seems like most of this is copied over from `SqlValidatorImpl#validateInsert` ; extended / refactored /  methods were extracted / commented on...
I think it might be challenging to maintain this in the long run - however `validateInsert` doesn't seem to be changing very often
I think it might be usefull to leave some comments about the origins of this method as an apidoc of `validateInsert` method